Sources/Workspace/Workspace.swift

```diff
@@ -224,6 +224,51 @@ public class Workspace {
             return lookup(package: name)?.manifest
         }
 
+        /// Returns constraints based on the dependencies.
+        ///
+        /// Versioned constraints are not added for dependencies present in the pins store.
+        func createConstraints(pinsStore: PinsStore) -> [RepositoryPackageConstraint] {
+            var constraints: [RepositoryPackageConstraint] = []
+            for (externalManifest, managedDependency) in dependencies {
+                let specifier = RepositorySpecifier(url: externalManifest.url)
+
+                if managedDependency.isInEditableState {
+                    // Create unversioned constraints for editable dependencies.
+                    let dependencies = externalManifest.package.dependencies.map{
+                        RepositoryPackageConstraint(
+                            container: RepositorySpecifier(url: $0.url),
+                            versionRequirement: .range($0.versionRange.asUtilityVersion))
+                    }
+
+                    constraints.append(
+                        RepositoryPackageConstraint(
+                            container: specifier, requirement: .unversioned(dependencies)))
+
+                } else if let version = managedDependency.currentVersion {
+                    // If this specifier is pinned, we should already have it
+                    // checked out at that version. We also don't need to add
+                    // the constraint again.
+                    if let pin = pinsStore.pinsMap[externalManifest.name] {
+                        assert(version == pin.version)
+                        continue
+                    }
+                    // If we know the manifest is at a particular version, use that.
+                    //
+                    // FIXME: This backfires in certain cases when the
+                    // graph is resolvable but this constraint makes the
+                    // resolution unsatisfiable.
+                    constraints.append(
+                        RepositoryPackageConstraint(
+                            container: specifier, versionRequirement: .exact(version)))
+                } else {
+                    // FIXME: Otherwise, we need to be able to constraint
+                    // precisely to the revision we have.
+                    fatalError("FIXME: Unimplemented.")
```

Comment:
Nope reaching this right now would mean a programming error in SwiftPM right now. This will be fixed soon when adding branch support I think.